workshop1/slides/src/slides.md

```diff
@@ -0,0 +1,664 @@
+
+![](images/data-transformation.pdf)
+
+
+## Writing `parseLine :: CSVLine -> Maybe Candidate`
+
+```haskell
+type CSVLine = String -- type alias
+
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (undefined)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate fields = undefined
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate maybeAge = undefined
+```
+
+* `_` means "for every other value"
+* `splitOn` splits a string into a list of strings on a character
+* `readMaybe` decodes a double from a string (may fail)
+
+
+## Writing `parseLine :: CSVLine -> Maybe Candidate`
+
+Our solution:
+
+```haskell
+parseLine :: CSVLine -> Maybe Candidate
+parseLine line = makeCandidate (splitOn ',' line)
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  fromMaybeAge (Candidate c p s) (readMaybe a)
+makeCandidate _ = Nothing
+
+fromMaybeAge
+  :: (Double -> Candidate)
+  -> Maybe Double
+  -> Maybe Candidate
+fromMaybeAge partialCandidate Nothing  = Nothing
+fromMaybeAge partialCandidate (Just a) = Just (partialCandidate a)
+```
+
+
+## Stripping Quotes
+
+```haskell
+stripQ :: String -> String
+stripQ = undefined
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = undefined
+```
+
+* Use `leftStrip` and `reverse` to implement `rightStrip`
+* Use `leftStrip` and `rightStrip` to implement `stripQ`
+* `reverse` reverses items in a list
+* `where` used to define local functions
+
+## Stripping Quotes
+
+Our solution:
+```haskell
+stripQ :: String -> String
+stripQ = leftStrip . rightStrip
+  where
+    leftStrip :: String -> String
+    leftStrip ('"' : xs) = xs
+    leftStrip xs         = xs
+
+    rightStrip :: String -> String
+    rightStrip = reverse . leftStrip . reverse
+```
+
+* Point-free style: composition FTW!
+
+## Improving `parseLine`
+
+* Using `stripQ` to improve `makeCandidate`
+
+```haskell
+...
+
+makeCandidate :: [String] -> Maybe Candidate
+makeCandidate [c, p, s, a] =
+  let c' = stripQ c
+      p' = stripQ p
+      s' = stripQ s
+  in fromMaybeAge (Candidate c' p' s') (readMaybe a)
+makeCandidate _ = Nothing
+
+...
+```
+
+* `let ... in ...` binding useful to split up functions
+
+
+## The `map` function
+
+* How do we apply `parseLine` to every row?
+* Use `map`, a *higher-order function*
+  * Takes two inputs: a function and a list.
+  * Applies the function to every item in the list
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+map f []     = []
+map f (x:xs) = f x : map f xs
+```
+
+* Defined using *recursion*
+
+
+## Writing `parseFile :: CSV -> DataModel`
+
+* We are ready to write the full parser function:
+
+```haskell
+type CSV = String
+type DataModel = [Maybe Candidate]
+
+parseFile :: CSV -> DataModel
+parseFile = undefined
+```
+
+* `lines` splits strings into lines
```

Comment:
This is such a simple function that I would almost point them to the cheat sheet first and have them find the functions needed. *Should* be simple enough since our cheat sheet is only an A4

Reply:
Fair